plugins/eos/gs-plugin-eos.c

```diff
@@ -267,6 +302,243 @@ read_icon_replacement_overrides (GHashTable *replacement_app_lookup)
+		guint64 total_size = 0;
+		if (gs_app_get_state (app) == AS_APP_STATE_INSTALLING) {
+			/* when the OS upgrade is already being installed and the
+			 * updater reports an available update, (meaning we were
+			 * polling before), we should readily call fetch */
+			os_upgrade_force_fetch (priv->updater_proxy);
+			break;
+		}
+
+		total_size = eos_updater_get_download_size (priv->updater_proxy);
+		gs_app_set_size_download (app, total_size);
+
+		gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+
+		break;
+	}
+	case EOS_UPDATER_STATE_ERROR: {
+		const gchar *error_name;
+		const gchar *error_message;
+		g_autoptr(GError) local_error = NULL;
+
+		error_name = eos_updater_get_error_name (priv->updater_proxy);
+		error_message = eos_updater_get_error_message (priv->updater_proxy);
+		local_error = g_dbus_error_new_for_dbus_error (error_name, error_message);
+
+		g_debug ("Got OS upgrade error state with name '%s': %s",
+			 error_name, error_message);
+
+		gs_app_set_state_recover (app);
+
+		/* only set up an error to be shown to the user if the user had
+		 * manually started the upgrade, and if the error in question is not
+		 * originated by the user canceling the upgrade */
+		if (os_upgrade_get_download_by_user (app) &&
+		    g_strcmp0 (error_name, "com.endlessm.Updater.Error.Cancelled") != 0) {
+			g_autoptr(GsPluginEvent) event = gs_plugin_event_new ();
+			gs_utils_error_convert_gdbus (&local_error);
+			gs_plugin_event_set_app (event, app);
+			gs_plugin_event_set_error (event, local_error);
+			gs_plugin_event_add_flag (event, GS_PLUGIN_EVENT_FLAG_WARNING);
+			gs_plugin_report_event (plugin, event);
+		}
+
+		break;
+	}
+	case EOS_UPDATER_STATE_FETCHING: {
+		guint64 total_size = 0;
+		guint64 downloaded = 0;
+		gfloat progress = 0;
+
+		/* ensure the state transition to 'installing' is allowed */
+		if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING)
+			gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+		gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+
+		downloaded = eos_updater_get_downloaded_bytes (priv->updater_proxy);
+		total_size = eos_updater_get_download_size (priv->updater_proxy);
+
+		if (total_size == 0) {
+			g_debug ("OS upgrade %s total size is 0!",
+				 gs_app_get_unique_id (app));
+		} else {
+			/* set progress only up to 90%, leaving the remaining
+			 * for applying the update */
+			progress = (gfloat) downloaded / (gfloat) total_size * 90.0;
+		}
+		gs_app_set_progress (app, (guint) progress);
+
+		break;
+	}
+	case EOS_UPDATER_STATE_UPDATE_READY: {
+		/* if there's an update ready to be deployed and we are not yet
+		 * showing the OS upgrade as downloading (state 'installing'), then
+		 * just show it as available so the user has a change to click 'download'
+		 * which will end up deploying the update */
+		if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING) {
+			gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+			break;
+		}
+
+		gs_app_set_progress (app, 95);
```

Comment:
I'd also prefer if this was in an `else` block and there was just a single `break`

Reply:
But do we also need to set the progress here? Is it not enough to do it later in the `APPLYING_UPDATE` transition?

Reply:
This was more like ensuring we are in that progress than anything, but yeah, I think it's safe to remove it.